src/javascript/tests/custom_types.rs

```diff
@@ -0,0 +1,153 @@
+use crate::assert_js;
+
+#[test]
+fn zero_arity_custom_type() {
+    assert_js!(
+        r#"
+type Mine{
+    This
+    ThatOneIsAMuchMuchMuchMuchMuchMuchMuchMuchMuchMuchMuchMuchLongerVariant
+}
+
+const this = This;
+const that = ThatOneIsAMuchMuchMuchMuchMuchMuchMuchMuchMuchMuchMuchMuchLongerVariant;
+
+fn go() {
+    This
+    ThatOneIsAMuchMuchMuchMuchMuchMuchMuchMuchMuchMuchMuchMuchLongerVariant
+}
+"#,
+        r#""use strict";
+
+const this = { type: "This" };
+
+const that = {
+  type: "ThatOneIsAMuchMuchMuchMuchMuchMuchMuchMuchMuchMuchMuchMuchLongerVariant"
+};
+
+function go() {
+  { type: "This" };
+  return {
+    type: "ThatOneIsAMuchMuchMuchMuchMuchMuchMuchMuchMuchMuchMuchMuchLongerVariant"
+  };
+}
+"#
+    );
+}
+
+#[test]
+fn custom_type_with_unnamed_fields() {
+    assert_js!(
+        r#"
+type Ip{
+    Ip(String)
+}
+
+const local = Ip("0.0.0.0")
+
+fn build(x) {
+    x("1.2.3.4")
+}
+
+fn go() {
+    build(Ip)
+    Ip("5.6.7.8")
+}
+
+// I don't think this accessor syntax is valid
+// fn access(ip: Ip) {
+//     ip.0
+// }
+"#,
+        r#""use strict";
+
+const local = { type: "Ip", 0: "0.0.0.0" };
+
+function build(x) {
+  return x("1.2.3.4");
+}
+
+function go() {
+  build((var0) => { return { type: "Ip", 0: var0 }; });
+  return { type: "Ip", 0: "5.6.7.8" };
+}
+"#
+    );
+
+    assert_js!(
+        r#"
+type TypeWithALongNameAndSeveralArguments{
+    TypeWithALongNameAndSeveralArguments(String, String, String, String, String)
+}
+
+
+fn go() {
+    TypeWithALongNameAndSeveralArguments
+    // TypeWithALongNameAndSeveralArguments("foo", "bar", "XXXXXXXXXXXXXXXXXXX", "baz", "last")
+}
+"#,
+        r#""use strict";
+
+function go() {
+  return (var0, var1, var2, var3, var4) => {
+    return {
+      type: "TypeWithALongNameAndSeveralArguments",
+      0: var0,
+      1: var1,
+      2: var2,
+      3: var3,
+      4: var4
+    };
+  };
+}
+"#
+    );
+}
+
+#[test]
+// How to we track to object keys, would we rather generate a function and call in order.
```

Comment:
What does this mean?

Reply:
Ummm,  I think that's very old musings will remove